proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2290 @@
+annotation are linear).
+
+Record patterns act like tuple patterns, but some fields can be
+omitted. A field can be omitted only if the resolved multiplicity for
+this field is ω.
+
+::
+
+  foo :: R' ->. A
+  foo {f2=x, f3=y} = … -- permitted as f1 has multiplicity ω
+  foo {f2=x} = … -- rejected as f3 is omitted and has multiplicity 1
+
+  foo :: R' -> A -- non-linear function!
+  foo {f2=x} = … -- permitted because the context has multiplicity ω,
+                 -- hence the resolved multiplicity of f3 is ω.
+
+Projections take an *unrestricted* record as argument: ``f1 :: R ->
+A1`` (because otherwise the other fields would not be consumed). There
+is an exception to this rule: if a record type has a single
+constructor, and all the other fields are unrestricted, then ``f1`` is
+made linear: ``f1 :: R ->. A1``. This non-uniformity is justified by
+the standard ``newtype`` idiom:
+
+::
+
+  newtype Foo = Foo { unFoo :: A }
+
+which becomes much less useful in linear code if ``unFoo :: Foo ->
+A``. Our practice of linear Haskell code indicates that this feature,
+while a mere convenience, is desirable (see *e.g.* `here
+<https://github.com/tweag/linear-base/blob/e72d996b5d0600b2d5f2483b95b064d524c83e46/src/System/IO/Resource.hs#L59-L61>`_).
+
+Records in GADT syntax
+~~~~~~~~~~~~~~~~~~~~~~
+
+Records can also be defined in GADT syntax:
+
+::
+
+  data R where
+    R :: { f1 :: A, f2 :: B } -> R
+
+In this special form, only the standard arrow is allowed, even with
+``-XLinearTypes``. This arrow, however, is not to be interpreted as
+the unrestricted arrow, or to have any meaning: it is just a syntactic
+construct. The multiplicity of the fields is given by the annotation
+on the binders, as with regular records.
+
+That is, in the above example, ``R`` has type
+
+::
+
+  R :: A ->. B ->. R
+
+In general, in
+
+::
+
+  data R where
+    R :: { f1 ::(π) A, f2 ::(ρ) B } -> R
+
+We have
+
+::
+
+  R :: A :π-> B :ρ-> R
+
+With absence of annotation interpreted as annotating with ``'One``.
+
+
+Inference
+---------
+
+.. _Inference:
+
+Because of backwards compatibility, we initially chose the following
+strategy: when the type of a function is not constrained by given
+constraints, we conservatively assume it to have multiplicity ω.
+
+Experience shows that this sometimes yield very confusing error messages
+where perfectly valid code is rejected:
+
+::
+
+  type family L x
+  type instance L Int = A ->. A
+
+  f :: L x -> x
+
+  u :: Int
+  u = f (\x -> x)
+
+While the identity function is indeed linear, because the resolution
+of the type family (``L Int ~ Int``) is delayed in GHC, ``\x -> x`` is
+considered to have no given type, and is inferred to have a non-linear
+type, and is refused by the type-checker.
+
+We therefore need a more refined strategy, to avoid surprising
+behaviour like the above. We do not expect it to be too hard to
+implement a better strategy, but we don't have a specification yet.
```

Comment:
This dangling unsolved question looms large for me. It would be helpful to have more information here -- at least in the form of several examples of when you expect GHC would infer linear arrows. Or maybe some guarantee of when GHC would _not_ infer linear arrows.